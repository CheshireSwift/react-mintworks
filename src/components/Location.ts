import { Fragment } from 'react';
import { div, h1, h, strong } from 'react-hyperscript-helpers';
import * as R from 'ramda';
import emoji from 'node-emoji';

import renderEffect from './renderEffect';

export enum LocationType {
  Basic,
  Advanced,
  Deed
}

type SpacesProps = {
  cost: string;
  count: number;
  filled: number;
};

export type LocationProps = {
  name: string;
  type: LocationType;
  spaces: SpacesProps;
  effect: string;
  owner?: string;
  ownerEffect?: string;
};

const typeIcons: { [T in LocationType]: string } = {
  [LocationType.Basic]: emoji.get('amphora'),
  [LocationType.Advanced]: emoji.get('zap'),
  [LocationType.Deed]: emoji.get('scroll')
};

const Space = (props: { filled: boolean; cost: string }) =>
  div(
    {
      style: {
        border: '2px dashed floralwhite',
        borderRadius: 9999,
        width: '2.5rem',
        height: '2.5rem',
        margin: 'auto',
        fontSize: 30
      }
    },
    props.filled ? emoji.get('white_circle') : props.cost
  );

const Spaces = ({
  available,
  count,
  cost,
  filled
}: SpacesProps & { available: boolean }) =>
  h(
    Fragment,
    available
      ? R.range(0, count).map(i => h(Space, { cost, filled: i < filled }))
      : ['Build deed to open']
  );

const Owner = (props: { owner?: string; effect: string }) =>
  div(
    {
      style: {
        backgroundColor: 'lightskyblue',
        position: 'absolute',
        bottom: '0.3rem',
        left: '0.3rem',
        right: '0.3rem',
        padding: '0.5em',
        borderRadius: 4
      }
    },
    [
      div(
        {
          style: {
            position: 'absolute',
            right: 0,
            bottom: '2rem',
            backgroundColor: 'lightskyblue',
            borderRadius: 4,
            padding: '0.25rem 1rem'
          }
        },
        [strong('Owner'), props.owner && ` (${props.owner})`]
      ),
      div(renderEffect(props.effect))
    ]
  );

const sectionStyle = {
  borderRadius: 4,
  backgroundColor: 'paleturquoise'
};

const cardStyle = {
  width: '32rem',
  borderRadius: 4,
  border: '1px solid black',
  backgroundColor: 'floralwhite',
  display: 'grid',
  gridTemplateAreas: `"spaces name"
        "spaces effect"`,
  gridTemplateColumns: '5rem auto',
  gridTemplateRows: '4rem 8rem',
  gridGap: '1rem',
  textAlign: 'center',
  padding: '1rem',
  fontSize: 'larger'
};

export const Location = ({
  name,
  type,
  effect,
  spaces,
  owner,
  ownerEffect
}: LocationProps) =>
  div({ style: cardStyle }, [
    div({ style: { ...sectionStyle, gridArea: 'name' } }, [
      h1({ style: { margin: '0.5rem' } }, name)
    ]),
    div(
      {
        style: {
          ...sectionStyle,
          background: 'transparent',
          position: 'absolute',
          top: '1.15rem',
          left: '6.2rem',
          fontSize: 48
        }
      },
      typeIcons[type]
    ),
    div(
      {
        style: {
          ...sectionStyle,
          gridArea: 'effect',
          paddingTop: '1rem',
          position: 'relative'
        }
      },
      [
        div(renderEffect(effect)),
        ownerEffect && h(Owner, { owner, effect: ownerEffect })
      ]
    ),
    div(
      {
        style: {
          ...sectionStyle,
          gridArea: 'spaces',
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-around',
          textTransform: 'uppercase',
          fontWeight: 'bold',
          fontSize: 24
        }
      },
      [h(Spaces, { ...spaces, available: !ownerEffect || !!owner })]
    )
  ]);
