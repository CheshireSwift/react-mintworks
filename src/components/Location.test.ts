import { render } from 'enzyme';
import { h } from 'react-hyperscript-helpers';
import emoji from 'node-emoji';

import { Location, LocationType } from './Location';

const mintIcon = emoji.get('white_circle');
const deedIcon = emoji.get('scroll');

describe('Location component', () => {
  const builderConfig = {
    name: 'Builder',
    type: LocationType.Basic,
    spaces: { available: true, cost: '2', count: 3, filled: 1 },
    effect: 'Build one of your Plans'
  };

  const wholesalerConfig = {
    name: 'Wholesaler',
    type: LocationType.Deed,
    spaces: { available: false, cost: '1', count: 1, filled: 0 },
    effect: 'Gain ()()',
    ownerEffect: '*Upkeep:* If occupied, Gain ()'
  };

  it('matches the builder snapshot', () => {
    expect(render(h(Location, builderConfig))).toMatchSnapshot();
  });

  it('matches the lotto snapshot', () => {
    expect(render(h(Location, wholesalerConfig))).toMatchSnapshot();
  });

  describe('basic display', () => {
    const builderText = render(h(Location, builderConfig)).text();

    it('includes the name', () => {
      expect(builderText).toContain(builderConfig.name);
    });

    it('includes the effect', () => {
      expect(builderText).toContain(builderConfig.effect);
    });

    it('includes the spaces', () => {
      expect(builderText).toContain(mintIcon + '2' + '2');
    });
  });

  describe('advanced display', () => {
    const wholesalerText = render(h(Location, wholesalerConfig)).text();

    it('includes the type icon', () => {
      expect(wholesalerText).toContain(deedIcon);
    });

    it('parses the effect text', () => {
      expect(wholesalerText).toContain('Gain ' + mintIcon + mintIcon);
    });

    it('includes the parsed owner text', () => {
      expect(wholesalerText).toContain('Owner');
      expect(wholesalerText).toContain('Upkeep: If occupied, Gain ' + mintIcon);
    });

    it('replaces the spaces with deed instructions when not available', () => {
      expect(wholesalerText).not.toContain('1');
      expect(wholesalerText).toContain('Build deed to open');
    });

    it('can display the owner when provided', () => {
      expect(
        render(h(Location, { ...wholesalerConfig, owner: 'Lisa' })).text()
      ).toContain('Lisa');
    });
  });
});
