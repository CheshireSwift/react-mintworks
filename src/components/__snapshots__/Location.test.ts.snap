// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location component matches the builder snapshot 1`] = `
<div
  style="width:32rem;border-radius:4px;border:1px solid black;background-color:floralwhite;display:grid;grid-template-areas:\\"spaces name\\"
        \\"spaces effect\\";grid-template-columns:5rem auto;grid-template-rows:4rem 8rem;grid-gap:1rem;text-align:center;padding:1rem;font-size:larger"
>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:name"
  >
    <h1
      style="margin:0.5rem"
    >
      Builder
    </h1>
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;background:transparent;position:absolute;top:1.15rem;left:6.2rem;font-size:48px"
  >
    🏺
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:effect;padding-top:1rem;position:relative"
  >
    <div>
      Build one of your Plans
    </div>
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:spaces;display:flex;flex-direction:column;justify-content:space-around;text-transform:uppercase;font-weight:bold;font-size:24px"
  >
    <div
      style="border:2px dashed floralwhite;border-radius:9999px;width:2.5rem;height:2.5rem;margin:auto;font-size:30px"
    >
      ⚪
    </div>
    <div
      style="border:2px dashed floralwhite;border-radius:9999px;width:2.5rem;height:2.5rem;margin:auto;font-size:30px"
    >
      2
    </div>
    <div
      style="border:2px dashed floralwhite;border-radius:9999px;width:2.5rem;height:2.5rem;margin:auto;font-size:30px"
    >
      2
    </div>
  </div>
</div>
`;

exports[`Location component matches the lotto snapshot 1`] = `
<div
  style="width:32rem;border-radius:4px;border:1px solid black;background-color:floralwhite;display:grid;grid-template-areas:\\"spaces name\\"
        \\"spaces effect\\";grid-template-columns:5rem auto;grid-template-rows:4rem 8rem;grid-gap:1rem;text-align:center;padding:1rem;font-size:larger"
>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:name"
  >
    <h1
      style="margin:0.5rem"
    >
      Wholesaler
    </h1>
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;background:transparent;position:absolute;top:1.15rem;left:6.2rem;font-size:48px"
  >
    📜
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:effect;padding-top:1rem;position:relative"
  >
    <div>
      Gain ⚪⚪
    </div>
    <div
      style="background-color:lightskyblue;position:absolute;bottom:0.3rem;left:0.3rem;right:0.3rem;padding:0.5em;border-radius:4px"
    >
      <div
        style="position:absolute;right:0;bottom:2rem;background-color:lightskyblue;border-radius:4px;padding:0.25rem 1rem"
      >
        <strong>
          Owner
        </strong>
      </div>
      <div>
        <strong>
          Upkeep:
        </strong>
         If occupied, Gain ⚪
      </div>
    </div>
  </div>
  <div
    style="border-radius:4px;background-color:paleturquoise;grid-area:spaces;display:flex;flex-direction:column;justify-content:space-around;text-transform:uppercase;font-weight:bold;font-size:24px"
  >
    Build deed to open
  </div>
</div>
`;
